File: app/server/middleware/rateLimit.js
/**
 * @fileoverview 速率限制中间件
 */

const { getClientIP } = require('../utils/ip');
const { RateLimitError } = require('../utils/errors');
const config = require('../utils/config');

/** @type {Map<string, {count: number, resetTime: number}>} */
const rateLimitMap = new Map();

/**
 * 速率限制中间件
 * @param {import('express').Request} req - Express请求对象
 * @param {import('express').Response} res - Express响应对象
 * @param {import('express').NextFunction} next - 下一个中间件
 */
function rateLimit(req, res, next) {
    const ip = getClientIP(req);
    const now = Date.now();
    
    const record = rateLimitMap.get(ip) || { count: 0, resetTime: now + config.rateLimit.windowMs };
    
    if (now > record.resetTime) {
        record.count = 1;
        record.resetTime = now + config.rateLimit.windowMs;
    } else {
        record.count++;
    }
    
    rateLimitMap.set(ip, record);
    
    if (record.count > config.rateLimit.maxRequests) {
        return next(new RateLimitError('请求过于频繁，请稍后再试'));
    }
    
    next();
}

/**
 * 登录速率限制
 * @type {Map<string, {count: number, lockoutUntil: number}>}
 */
const loginAttemptsMap = new Map();

/**
 * 获取登录尝试记录
 * @param {string} ip - IP地址
 * @returns {{count: number, lockoutUntil: number}}
 */
function getLoginAttempts(ip) {
    const attempts = loginAttemptsMap.get(ip) || { count: 0, lockoutUntil: 0 };
    if (attempts.lockoutUntil && Date.now() > attempts.lockoutUntil) {
        loginAttemptsMap.delete(ip);
        return { count: 0, lockoutUntil: 0 };
    }
    return attempts;
}

/**
 * 记录登录尝试
 * @param {string} ip - IP地址
 * @param {boolean} success - 是否成功
 */
function recordLoginAttempt(ip, success) {
    if (success) {
        loginAttemptsMap.delete(ip);
        return;
    }
    
    const attempts = getLoginAttempts(ip);
    attempts.count++;
    if (attempts.count >= config.login.maxAttempts) {
        attempts.lockoutUntil = Date.now() + config.login.lockoutTime;
    }
    loginAttemptsMap.set(ip, attempts);
}

/**
 * 检查是否被锁定
 * @param {string} ip - IP地址
 * @returns {boolean}
 */
function isLockedOut(ip) {
    const attempts = getLoginAttempts(ip);
    return attempts.lockoutUntil > 0;
}

/**
 * 获取剩余尝试次数
 * @param {string} ip - IP地址
 * @returns {number}
 */
function getRemainingAttempts(ip) {
    const attempts = getLoginAttempts(ip);
    return Math.max(0, config.login.maxAttempts - attempts.count);
}

module.exports = {
    rateLimit,
    getLoginAttempts,
    recordLoginAttempt,
    isLockedOut,
    getRemainingAttempts
};

File: app/server/internal/middleware/ratelimit.go
package middleware

import (
	"strconv"
	"sync"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/sushazhi/fnos-logmanager/internal/config"
	apperrors "github.com/sushazhi/fnos-logmanager/internal/errors"
	"github.com/sushazhi/fnos-logmanager/internal/types"
	"github.com/sushazhi/fnos-logmanager/internal/utils"
)

const maxRateLimitEntries = 10000

type rateLimitStore struct {
	mu    sync.RWMutex
	store map[string]*types.RateLimitRecord
	// oldestKey tracks the oldest entry for deterministic LRU eviction.
	oldestKey  string
	oldestTime int64
}

func newRateLimitStore() *rateLimitStore {
	return &rateLimitStore{
		store: make(map[string]*types.RateLimitRecord),
	}
}

// bump atomically increments the counter for key, resetting it when the window
// has elapsed, and returns a snapshot of the resulting record. The whole
// read-modify-write happens under a single write lock, which eliminates the
// data race that occurred when middlewares incremented a shared *Record that
// had been returned by an earlier read-locked lookup.
func (rls *rateLimitStore) bump(key string, now, windowMs int64) *types.RateLimitRecord {
	rls.mu.Lock()
	defer rls.mu.Unlock()

	rec := rls.store[key]
	if rec == nil {
		rec = &types.RateLimitRecord{
			Count:     0,
			ResetTime: now + windowMs,
		}
	}

	if now > rec.ResetTime {
		rec.Count = 1
		rec.ResetTime = now + windowMs
	} else {
		rec.Count++
	}

	if len(rls.store) >= maxRateLimitEntries {
		// Evict the oldest entry deterministically instead of random.
		// If the tracked oldest key is stale, scan for the real oldest.
		if rls.oldestKey == "" || rls.store[rls.oldestKey] == nil {
			var minTime int64 = 1<<63 - 1
			var minKey string
			for k, v := range rls.store {
				if v.ResetTime < minTime {
					minTime = v.ResetTime
					minKey = k
				}
			}
			rls.oldestKey = minKey
			rls.oldestTime = minTime
		}
		if rls.oldestKey != "" {
			delete(rls.store, rls.oldestKey)
			rls.oldestKey = ""
		}
	}
	// Track the new oldest entry
	if rls.oldestKey == "" || rec.ResetTime < rls.oldestTime {
		rls.oldestTime = rec.ResetTime
	}

	rls.store[key] = rec
	cp := *rec
	return &cp
}

func (rls *rateLimitStore) cleanup() {
	rls.mu.Lock()
	defer rls.mu.Unlock()
	now := time.Now().UnixMilli()
	for key, record := range rls.store {
		if now > record.ResetTime {
			delete(rls.store, key)
		}
	}
}

var (
	globalRateLimit    = newRateLimitStore()
	apiRateLimitStore  = newRateLimitStore()
	sensitiveRateLimit = newRateLimitStore()
)

func init() {
	go func() {
		ticker := time.NewTicker(60 * time.Second)
		for range ticker.C {
			globalRateLimit.cleanup()
			apiRateLimitStore.cleanup()
			sensitiveRateLimit.cleanup()
		}
	}()
}

// RateLimit is the global rate limiting middleware.
func RateLimit(c *gin.Context) {
	cfg := config.Get()
	ip := utils.GetClientIP(c.Request)
	now := time.Now().UnixMilli()

	record := globalRateLimit.bump(ip, now, cfg.RateLimit.WindowMs)

	c.Header("X-RateLimit-Limit", strconv.Itoa(cfg.RateLimit.MaxRequests))
	c.Header("X-RateLimit-Remaining", strconv.Itoa(max(0, cfg.RateLimit.MaxRequests-record.Count)))
	c.Header("X-RateLimit-Reset", strconv.FormatInt(record.ResetTime, 10))

	if record.Count > cfg.RateLimit.MaxRequests {
		c.Header("Retry-After", strconv.Itoa(int((record.ResetTime-now)/1000)))
		c.Error(apperrors.NewRateLimitError("请求过于频繁，请稍后再试"))
		c.Abort()
		return
	}

	c.Next()
}

// APIRateLimit returns a middleware that applies per-endpoint rate limiting.
func APIRateLimit(maxRequests int, windowMs int64) gin.HandlerFunc {
	return func(c *gin.Context) {
		ip := utils.GetClientIP(c.Request)
		now := time.Now().UnixMilli()

		record := apiRateLimitStore.bump(ip, now, windowMs)

		if record.Count > maxRequests {
			c.Header("Retry-After", strconv.Itoa(int((record.ResetTime-now)/1000)))
			c.Error(apperrors.NewRateLimitError("API请求过于频繁，请稍后再试"))
			c.Abort()
			return
		}

		c.Next()
	}
}

// SensitiveActionRateLimit returns a middleware that limits sensitive operations per IP+path.
func SensitiveActionRateLimit(maxRequests int, windowMs int64) gin.HandlerFunc {
	return func(c *gin.Context) {
		ip := utils.GetClientIP(c.Request)
		now := time.Now().UnixMilli()
		key := ip + ":" + c.Request.URL.Path

		record := sensitiveRateLimit.bump(key, now, windowMs)

		if record.Count > maxRequests {
			c.Header("Retry-After", strconv.Itoa(int((record.ResetTime-now)/1000)))
			c.Error(apperrors.NewRateLimitError("敏感操作过于频繁，请稍后再试"))
			c.Abort()
			return
		}

		c.Next()
	}
}
